Move sys-mor solver glue code into pymor.bindings

sdrave opened the pull request:
Hi @pmli, as discussed in #291, I have moved in this pull request all matrix equation solver glue code into the `bindings` package. The solver dispatch is now handled by `LTISystem._lyap_solver` and `LTISystem._ricc_solver`. Moreover, I have made `me_solver` an attribute of the `LTISystem` with configurable `default`, so that the argument does not have to be passed around everywhere.

Please let me know what you think! (I had to merge with master, so only the last two commits are really new ...)

sdrave commented:
BTW, if you agree, I would like to add the `tol` parameter to the `solver_options` as well ...

pmli commented on src/pymor/discretizations/iosys.py:
```diff
@@ -157,6 +157,11 @@ def mag_plot(cls, sys_list, plot_style_list=None, w=None, ord=None, dB=False, Hz
 
 
+_DEFAULT_ME_SOLVER = 'slycot' if config.HAVE_SLYCOT else \
+                     'pymess' if config.HAVE_PYMESS else \
```
`pymess` should be prefered over `slycot`, since `slycot` can't handle large equations.

pmli commented on src/pymor/discretizations/iosys.py:
```diff
@@ -466,8 +474,32 @@ def eval_dtf(self, s):
+            from pymor.bindings.slycot import solve_lyap as solve_lyap_impl
+        elif solver.startswith('pymess'):
+            from pymor.bindings.pymess import solve_lyap as solve_lyap_impl
+            solve_lyap_impl = solve_lyap_impl.partial(me_solver=solver)
```
Is the indentation here too big?

pmli commented on src/pymor/discretizations/iosys.py:
```diff
@@ -466,8 +474,32 @@ def eval_dtf(self, s):
+            from pymor.bindings.slycot import solve_ricc as solve_ricc_impl
+        elif solver.startswith('pymess'):
+            from pymor.bindings.pymess import solve_ricc as solve_ricc_impl
+            solve_ricc_impl = solve_ricc_impl.partial(me_solver=solver)
```
See above.

pmli commented on src/pymor/discretizations/iosys.py:
```diff
@@ -511,46 +539,42 @@ def gramian(self, typ, subtyp, me_solver=None, tol=None):
             if subtyp == 'cf':
-                return solve_ricc(A, E=E, B=B * c, C=C * c, R=IdentityOperator(C.range) * (-1),
-                                  trans=True, me_solver=me_solver, tol=tol)
+                return self.ricc_solver()(A, E=E, B=B * c, C=C * c, R=IdentityOperator(C.range) * (-1),
```
Should there be one more underscore here? (`self._ricc_solver`)

pmli commented on src/pymor/discretizations/iosys.py:
```diff
@@ -511,46 +539,42 @@ def gramian(self, typ, subtyp, me_solver=None, tol=None):
             elif subtyp == 'of':
-                return solve_ricc(A, E=E, B=B * c, C=C * c, R=IdentityOperator(B.source) * (-1),
-                                  trans=False, me_solver=me_solver, tol=tol)
+                return self.ricc_solver()(A, E=E, B=B * c, C=C * c, R=IdentityOperator(B.source) * (-1),
```
See above.

pmli commented:
I made some minor comments. In general, I like bindings structure.

Yes, it would make sense to add `tol` to `solver_options`.

sdrave commented:
Hi @pmli, I have changed the default default backend to pymess, fixed the `_ricc_solver` calls and `me_solver / tol` have been replaced by `options` dicts everywhere. Maybe a bit over-engineered, but now we have the same interface as for linear solvers and the whole design is easily extensible.

Without pymess, I cannot test everything, but from my side this could be merged as is into the sys-mor branch.

pmli commented:
Tests in `pymortests` do not work, since they import from `pymor.algorithms`. I'll update them and see if things work with pymess.

pmli commented:
Now both Lyapunov and Riccati equation tests pass on my machine.

sdrave commented:
Hi @pmli, what's the status on this? Can we merge?

pmli commented:
Yes, IMO.
